plugins/destination/duckdb/client/write.go

```diff
@@ -134,39 +125,46 @@ func (c *Client) WriteTableBatch(ctx context.Context, table *schema.Table, recor
-		if err := c.createTableIfNotExist(ctx, tmpTableName, table); err != nil {
-			return fmt.Errorf("failed to create table %s: %w", tmpTableName, err)
-		}
-		if err := c.copyFromFile(ctx, tmpTableName, f.Name(), sc); err != nil {
-			return fmt.Errorf("failed to copy from file %s: %w", f.Name(), err)
-		}
+	if !c.enabledPks() || len(table.PrimaryKeys()) == 0 {
+		return c.copyFromFile(ctx, table.Name, f.Name(), sc)
+	}
 
-		// At time of writing (March 2023), duckdb does not support updating list columns.
-		// As a workaround, we delete the row and insert it again. This makes it non-atomic, unfortunately,
-		// but this is unavoidable until support is added to duckdb itself.
-		// See https://github.com/duckdb/duckdb/blob/c5d9afb97bbf0be12216f3b89ae3131afbbc3156/src/storage/table/list_column_data.cpp#L243-L251
-		if containsList(table) {
-			if err := c.deleteByPK(ctx, tmpTableName, table.Name, table); err != nil {
-				return err
-			}
-			if _, err = c.db.ExecContext(ctx, "insert into "+table.Name+" from "+tmpTableName); err != nil {
-				return fmt.Errorf("failed to insert into %s from %s: %w", table.Name, tmpTableName, err)
-			}
-		} else {
-			if err := c.upsert(ctx, tmpTableName, table.Name, table); err != nil {
-				return err
-			}
-		}
-		if _, err = c.db.ExecContext(ctx, "drop table "+tmpTableName); err != nil {
-			return err
+	tmpTableName := table.Name + strings.ReplaceAll(uuid.New().String(), "-", "_")
+	if err := c.createTableIfNotExist(ctx, tmpTableName, table); err != nil {
+		return fmt.Errorf("failed to create table %s: %w", tmpTableName, err)
+	}
+	defer func() {
+		e := c.exec(ctx, "drop table "+tmpTableName)
+		if err == nil {
+			// we preserve original error, so update only on nil err
+			err = e
 		}
+	}()
+	if err := c.copyFromFile(ctx, tmpTableName, f.Name(), sc); err != nil {
+		return fmt.Errorf("failed to copy from file %s: %w", f.Name(), err)
 	}
 
-	return nil
+	// At time of writing (March 2023), duckdb does not support updating list columns.
+	// As a workaround, we delete the row and insert it again. This makes it non-atomic, unfortunately,
+	// but this is unavoidable until support is added to duckdb itself.
+	// See https://github.com/duckdb/duckdb/blob/c5d9afb97bbf0be12216f3b89ae3131afbbc3156/src/storage/table/list_column_data.cpp#L243-L251
+	if containsList(table) {
+		return c.deleteInsert(ctx, tmpTableName, table)
+	}
+
+	return c.upsert(ctx, tmpTableName, table)
+}
+
+func (c *Client) deleteInsert(ctx context.Context, tmpTableName string, table *schema.Table) error {
+	if err := c.deleteByPK(ctx, tmpTableName, table); err != nil {
+		return err
+	}
+
+	// per https://duckdb.org/docs/sql/indexes#over-eager-unique-constraint-checking we might need to retry
+	return backoff.Retry(
+		func() error {
+			return c.exec(ctx, "insert into "+table.Name+" from "+tmpTableName)
+		},
+		backoff.WithContext(backoff.WithMaxRetries(backoff.NewExponentialBackOff(), 3), ctx),
```

Comment:
How long is the backoff period in total after 3 retries? Should we make the number of retries configurable? Because I think the number you'll need will depend on how many rows are being inserted / deleted

Reply:
I moved to a constant backoff with 50ms.
We can address this if someone reports issues even after this fix.

Reply:
It's hard to know whether this fix is going to generalize beyond the test, because we're only inserting a few rows there.  So I'd probably lean towards a higher number of retries, at least, if not making it fully configurable. But yeah we can see if we get any reports of this issue after the change.